build.gradle

```diff
@@ -42,39 +49,18 @@ ext {
     nhaarmanMockitoVersion = "2.2.0"
 }
 
-buildscript {
-    ext.kotlinVersion = '1.5.32'
-    ext.navComponentVersion = '2.3.5'
-    ext.detektVersion = '1.15.0'
+allprojects {
+    apply plugin: 'checkstyle'
 
-    repositories {
-        maven {
-            url 'https://a8c-libs.s3.amazonaws.com/android'
-            content {
-                includeGroup "com.automattic.android"
-            }
+    configurations.all {
+        resolutionStrategy {
+            // Don't allow Gradle to resolve the dependency to any other value
+            force "org.jetbrains.kotlin:kotlin-stdlib:$gradle.ext.kotlinVersion"
+            force "org.jetbrains.kotlin:kotlin-stdlib-common:$gradle.ext.kotlinVersion"
+            force "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$gradle.ext.kotlinVersion"
```

Comment:
Thanks for the ping @wzieba, as we discussed before, I agree we should keep the same version that the kotlin's plugin uses for the runtime dependencies as well.

> I believe it's best to leave Kotlin resolution as it is and accept that external plugins can influence version of Kotlin we use

I'm afraid this won't be very safe, kotlin's updates need proper testing to confirm there are no hidden regressions in different Android APIs, and if we accept updates coming from external plugins, they will go unnoticed, unless there is a compile-time regression, so I prefer to hold on updating the external dependency than updating kotlin's version.

I know we didn't find a way to enforce the plugin version, but I think we can at least make the build fail when a different version than what we expect is resolved, we can probably add some [constraints](https://docs.gradle.org/current/userguide/dependency_constraints.html#sec:adding-constraints-transitive-deps) using a `strictly` version, this would make the build fail if a transitive dependency caused an update implicitly, the code would be something like this:

```
    constraints {
        for (module in ["stdlib", "stdlib-jdk7", "stdlib-jdk8"]) {
            implementation("org.jetbrains.kotlin:kotlin-$module") {
                version {
                    strictly gradle.ext.kotlinVersion
                }
            }
        }
    }
```

WDYT about this? I tested this with my navigation's update PR, and this constraint causes the build to fail, while it succeeds when I downgrade the `Navigation` version

Reply:
Sorry for long reply

> I prefer to hold on updating the external dependency than updating kotlin's version.

Sure, my phrasing was not the best but what I meant is to accept the fact of this behaviour and not force our own standard library.

I like the way of handling you proposed. I didn't use constraints but as you say if:

> we can at least make the build fail when a different version than what we expect is resolved

it would be awesome!

Reply:
@wzieba I understand where you're coming from, but in practice I don't think it makes sense to blindly accept a transitive dependency version. There might be a case where it's better to update our project and there might be a case where we have to force our own version. Your initial argument doesn't answer recommend a way to address the failure as far as I can tell, so I couldn't compare the trade-offs very well.

> Unfortunately, we didn't find any way of forcing specific version of Kotlin Gradle Plugin and our bump is forced by other plugin (safeargs in this case).

@wzieba @hichamboushaba I think it's possible to force a specific version, let me try it out tomorrow and get back to you.

> I know we didn't find a way to enforce the plugin version, but I think we can at least make the build fail when a different version than what we expect is resolved, we can probably add some constraints using a strictly version, this would make the build fail if a transitive dependency caused an update implicitly, the code would be something like this:

As far as I can tell, this is just a different syntax that'll result in the same thing because `strictly` will also force the same resolution. Having said that, this seems like the better syntax. If we have to force the version, and assuming the constraints get us to the solution we want, I like using this. Thank you for sharing @hichamboushaba.

---

@wzieba @hichamboushaba The plugin version being possibly different was a very important point, thank you for bringing it up, I unfortunately missed this idea in my initial implementation. I think it's possible to add a constraint to the plugin version resolution and once that's done for Kotlin plugin, it might be possible to avoid the extra step by removing the direct dependencies as suggested by @wzieba [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15314#discussion_r775477847). I didn't want to make that change as part of this PR [as I explained in my reply](https://github.com/wordpress-mobile/WordPress-Android/pull/15314#discussion_r778482297), but I definitely don't mind doing it if there is an extra benefit to it.

Fwiw, I think the issue we are dealing is potentially a bug and if we can get away without using these extra constraints, we should remove it. (possibly after upgrading to Kotlin `1.6`) I still think it's a good idea to have more control over how our dependencies are resolved, but I'd prefer a more general solution than just forcing one dependency to be resolved in a certain way.

Reply:
> As far as I can tell, this is just a different syntax that'll result in the same thing because strictly will also force the same resolution. Having said that, this seems like the better syntax. If we have to force the version, and assuming the constraints get us to the solution we want, I like using this.

I'm not sure it's the same thing (but I may be wrong, my knowledge of gradle is very limited 🙂), the `resolutionStrategy` will force the dependency, so the build won't fail, even if a library is bringing a dependency with a higher version, where the `constraints` block will let Gradle do its work for resolving the version it needs, and it will then check it against the stated `constraints`, and the build will fail if they can't be respected.

>  I think it's possible to force a specific version, let me try it out tomorrow and get back to you.

This would be great if possible, thanks.

Reply:
@hichamboushaba @wzieba Just wanted to give you a heads up that it'll be a while before I can get back to this PR. I'm going to convert it back to draft and will ping you again when it's ready. Thank you for your feedback so far 🙇‍♂️